base/src/com/google/idea/blaze/base/command/info/BlazeInfoProvider.java

```diff
@@ -0,0 +1,124 @@
+ *
+ * @see BlazeInfoRunner
+ */
+public class BlazeInfoProvider {
+    private static final Logger logger = Logger.getInstance(BlazeInfoProvider.class);
+    private static final BoolExperiment enabled =
+            new BoolExperiment("blaze.info.provider.enabled", false);
+
+    private final Project project;
+    private volatile BlazeInfo blazeInfo;
+
+    public BlazeInfoProvider(Project project) {
+        this.project = project;
+    }
+
+    public static BlazeInfoProvider getInstance(Project project) {
+        return project.getService(BlazeInfoProvider.class);
+    }
+
+    public static Boolean isEnabled() {
+        return enabled.getValue();
+    }
+
+    public ListenableFuture<BlazeInfo> getBlazeInfo(
+            BlazeContext context,
+            List<String> blazeFlags) {
+        return BlazeExecutor.getInstance().submit(() -> {
+            BlazeInfo info = getCachedBlazeInfo(context,
+                    blazeFlags);
+            if (info == null) {
+                throw new BlazeInfoException("Unable to get blaze info");
+            }
+            return info;
+        });
+    }
+
+    private @Nullable BlazeInfo getCachedBlazeInfo(
+            BlazeContext context,
+            List<String> blazeFlags) {
+        if (blazeInfo != null) {
+            logger.debug("Returning cached BlazeInfo");
+            return blazeInfo;
+        }
+        try {
+            BlazeImportSettings importSettings = BlazeImportSettingsManager.getInstance(project).getImportSettings();
+            if (importSettings == null) {
+                return null;
+            }
+            BuildSystem.BuildInvoker buildInvoker = Blaze.getBuildSystemProvider(project)
+                    .getBuildSystem()
+                    .getDefaultInvoker(project, context);
+            logger.debug("Running bazel info");
+            blazeInfo = BlazeInfoRunner.getInstance().runBlazeInfo(
+                    project,
+                    buildInvoker,
+                    BlazeContext.create(),
+                    importSettings.getBuildSystem(),
+                    blazeFlags
+            ).get();
+            return blazeInfo;
+        } catch (InterruptedException | ExecutionException e) {
+            logger.warn("Unable to run blaze info", e);
+            return null;
+        }
+    }
+
+    public void invalidate() {
+        logger.debug("invalidating");
+        blazeInfo = null;
+    }
+
+    public static final class Invalidator implements SyncListener {
+        @Override
+        public void afterSync(
+                Project project,
+                BlazeContext context,
+                SyncMode syncMode,
+                SyncResult syncResult,
+                ImmutableSet<Integer> buildIds) {
+            if(!isEnabled()) {
+                return;
+            }
+            BlazeInfoProvider.getInstance(project).invalidate();
```

Comment:
Should we check for null here? In case we de-register the provider accidentally and forget to de-register the invalidator.